client/src/services/hotkeyManager.ts

```diff
@@ -0,0 +1,126 @@
+
+export const HotkeyManagerKey = 'hotkey';
+
+export enum Modifiers {
+  Ctrl = 1 << 0,
+  Alt = 1 << 1,
+  Shift = 1 << 2,
+}
+
+export enum Platforms {
+  Windows = 0,
+  MacOS = 1 << 0,
+  Linux = 1 << 1,
+  Web = 1 << 2,
+  Mobile = 1 << 3,
+  Desktop = Windows | MacOS | Linux,
+  All = Desktop | Web | Mobile,
+}
+
+export interface Hotkey {
+  key: string;
+  modifiers: Modifiers;
+  platforms: Platforms;
+  command: () => Promise<void>;
+}
+
+class Hotkeys {
+  currentHotkeys: Hotkey[];
+  id: number;
+
+  constructor(id: number) {
+    this.currentHotkeys = [];
+    this.id = id;
+  }
+
+  async add(key: string, modifiers: Modifiers, platforms: Platforms, command: () => Promise<void>): Promise<void> {
+    this.currentHotkeys.push({ key: key, modifiers: modifiers, platforms: platforms, command: command });
+  }
+}
+
+export class HotkeyManager {
+  currentHotkeys: Hotkeys[];
+  index: number;
+
+  constructor() {
+    this.currentHotkeys = [];
+    this.index = 0;
+    window.addEventListener('keydown', async (event: KeyboardEvent): Promise<void> => await this.onKeyPress(event, this.currentHotkeys));
+  }
+
+  newHotkeys = (): Hotkeys => {
+    const newKeys = new Hotkeys(this.index);
+    this.index++;
+    this.currentHotkeys.push(newKeys);
+    return newKeys;
+  };
+
+  async unregister(toRemove: Hotkeys): Promise<void> {
+    this.currentHotkeys = this.currentHotkeys.filter((hotkeys) => hotkeys.id !== toRemove.id);
+  }
+
+  private async onKeyPress(event: KeyboardEvent, hotKeys: Array<Hotkeys>): Promise<void> {
+    if (!isDesktop() && !isWeb()) {
+      return;
+    }
+    for (const keyPresses of hotKeys) {
+      for (const keyPress of keyPresses.currentHotkeys) {
+        if (!(await this.checkPlatforms(keyPress.platforms))) {
+          continue;
+        }
+        if (event.key.toLowerCase() === keyPress.key) {
+          if ((await this.checkModifiers(event, keyPress.modifiers)) === false) {
+            continue;
+          }
+          event.preventDefault();
+          await keyPress.command();
+        }
+      }
+    }
+  }
+
+  private async checkModifiers(event: KeyboardEvent, modifiers?: number): Promise<boolean> {
+    if (!modifiers) {
+      return !event.shiftKey && !event.ctrlKey && !event.metaKey && !event.altKey;
+    }
+    const ctrlKey = isMacOS() ? event.metaKey : isDesktop() ? event.ctrlKey : event.ctrlKey || event.metaKey;
+    switch (modifiers) {
+      case Modifiers.Ctrl:
+        return ctrlKey && !event.shiftKey && !event.altKey;
+      case Modifiers.Shift:
+        return !ctrlKey && event.shiftKey && !event.altKey;
+      case Modifiers.Alt:
+        return !ctrlKey && !event.shiftKey && event.altKey;
+      case Modifiers.Ctrl + Modifiers.Shift:
+        return ctrlKey && event.shiftKey && !event.altKey;
+      default:
+        return false;
+    }
+  }
+
+  private async checkPlatforms(platforms: Platforms): Promise<boolean> {
+    switch (platforms) {
```

Comment:
It's a bitmask https://en.wikipedia.org/wiki/Mask_(computing)#:~:text=In%20computer%20science%2C%20a%20mask,in%20a%20single%20bitwise%20operation.